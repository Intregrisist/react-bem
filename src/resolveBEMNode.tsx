import { BEMModifiers } from 'bem-helpers'
import * as React from 'react'

import { bemClassNameProp, isString, omitBEMProps } from './helpers'
import { ReactBEMElementProps } from './types'

/**
 * BEM block or element node.
 */
export type ReactBEMElement = React.ReactElement<ReactBEMElementProps>

/**
 * Walks a BEM node tree, consumes BEM props and resolves class names.
 */
export default function resolveBEMNode(
	node: ReactBEMElement,
	bem: {
		block: string,
		element?: string,
		modifiers?: BEMModifiers
	},
	/**
	 * Recursion depth.
	 */
	depth: number = 0,
): JSX.Element {
	if (!node) {
		throw new Error('Missing required React element')
	}
	if (!React.isValidElement(node)) {
		throw new Error('Invalid React element')
	}
	if (!bem) {
		throw new Error('Missing required BEM object argument')
	}
	if (!bem.block) {
		throw new Error('Missing required BEM block name')
	}
	if (!isString(bem.block)) {
		throw new Error('BEM block name must be a string')
	}
	if (bem.element && !isString(bem.element)) {
		throw new Error('BEM element name must be a string')
	}
	const props = omitBEMProps(node.props)
	const classNameProp = (bem.element || !depth) ? bemClassNameProp(
		bem.block,
		bem.element,
		bem.modifiers,
		{ className: props.className },
	) : {}
	return (
		<node.type
			{...{
				...props,
				children: resolveChildren(props.children, bem.block),
				...classNameProp,
			}}
		/>
	)

	function resolveChildren(children: React.ReactNode, block: string) {
		const count = React.Children.count(children)
		return !count ? children : (count > 1)
			? React.Children.map(children, resolveChild)
			: resolveChild(React.Children.only(children), 0, true)

		function resolveChild(
			child: ReactBEMElement,
			index: number,
			isOnlyChild: Boolean = false,
		) {
			const { element, modifiers } = child.props
			return (!element && !isOnlyChild) ? child : resolveBEMNode(
				child,
				{ block, element, modifiers },
				depth + 1,
			)
		}
	}
}
