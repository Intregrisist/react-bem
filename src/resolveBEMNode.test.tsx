import { shallow } from 'enzyme'
import * as React from 'react'

import resolveBEMNode from './resolveBEMNode'

describe('resolveBEMNode', () => {

	it('throws if missing React element argument', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode(undefined as any, undefined as any)
		}).toThrow('Missing required React element')
	})

	it('throws if invalid React argument is passed', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode('foo' as any, undefined as any)
		}).toThrow('Invalid React element')
	})

	it('throws if missing BEM object argument', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode(<div />, undefined as any)
		}).toThrow('Missing required BEM object argument')
	})

	it('throws if missing BEM block name', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode(<div />, {} as any)
		}).toThrow('Missing required BEM block name')
	})

	it('throws if BEM block name is a number', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode(<div />, { block: 42 } as any)
		}).toThrow('BEM block name must be a string')
	})

	it('throws if BEM element name is a number', () => {
		expect(() => {
			// tslint:disable-next-line:no-any
			resolveBEMNode(<div />, { block: 'foo', element: 42 } as any)
		}).toThrow('BEM element name must be a string')
	})

	it('assigns block name as class name', () => {
		expect(shallow(resolveBEMNode(
			<div />,
			{
				block: 'block',
			},
		))).toMatchSnapshot()
	})

	it('assigns "block__element" to root node\'s className', () => {
		expect(shallow(resolveBEMNode(
			<div />,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('preserves existing className', () => {
		expect(shallow(resolveBEMNode(
			<div element="remove-me" modifiers="and-me" className="keep-me" />,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('removes element and modifiers props from root node', () => {
		expect(shallow(resolveBEMNode(
			<div element="remove-me" modifiers="and-me" />,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('resolves a nested element', () => {
		expect(shallow(resolveBEMNode(
			<div>
				<div element="nested" />
			</div>,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('resolves a deeply nested element', () => {
		expect(shallow(resolveBEMNode(
			<div>
				<div>
					<div element="deeply-nested" />
				</div>
			</div>,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('removes an empty element value', () => {
		expect(shallow(resolveBEMNode(
			<div>
				<div element="" modifiers="mod1 mod2">
					<div element="deeply-nested" />
				</div>
			</div>,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})

	it('resolves a deeply nested element with modifiers', () => {
		expect(shallow(resolveBEMNode(
			<div>
				<div>
					<div element="deeply-nested" modifiers="mod1 mod2" />
				</div>
			</div>,
			{
				block: 'block',
				element: 'element',
			},
		))).toMatchSnapshot()
	})
})
