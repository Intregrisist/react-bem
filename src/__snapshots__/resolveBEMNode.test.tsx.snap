// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveBEMNode assigns "block__element" to root node's className 1`] = `
<div
  className="block__element"
/>
`;

exports[`resolveBEMNode assigns block name as class name 1`] = `
<div
  className="block"
/>
`;

exports[`resolveBEMNode doesn't choke on nested primitive values 1`] = `
<div
  className="block"
/>
`;

exports[`resolveBEMNode doesn't choke on nested primitive values 2`] = `
<div
  className="block"
/>
`;

exports[`resolveBEMNode doesn't choke on nested primitive values 3`] = `
<div
  className="block"
>
  foo
</div>
`;

exports[`resolveBEMNode doesn't choke on nested primitive values 4`] = `
<div
  className="block"
>
  42
</div>
`;

exports[`resolveBEMNode preserves existing className 1`] = `
<div
  className="block__element keep-me"
/>
`;

exports[`resolveBEMNode removes an empty element value 1`] = `
<div
  className="block__element"
>
  <div>
    <div
      className="block__deeply-nested"
    />
  </div>
</div>
`;

exports[`resolveBEMNode removes element and modifiers props from root node 1`] = `
<div
  className="block__element"
/>
`;

exports[`resolveBEMNode resolves a deeply nested element 1`] = `
<div
  className="block__element"
>
  <div>
    <div
      className="block__deeply-nested"
    />
  </div>
</div>
`;

exports[`resolveBEMNode resolves a deeply nested element with modifiers 1`] = `
<div
  className="block__element"
>
  <div>
    <div
      className="block__deeply-nested block__deeply-nested--mod1 block__deeply-nested--mod2"
    />
  </div>
</div>
`;

exports[`resolveBEMNode resolves a nested element 1`] = `
<div
  className="block__element"
>
  <div
    className="block__nested"
  />
</div>
`;
